libraries/chain/controller.cpp

```diff
@@ -3468,9 +3471,14 @@ void controller::validate_tapos( const transaction& trx )const { try {
 } FC_CAPTURE_AND_RETHROW() }
 
 void controller::validate_db_available_size() const {
-   const auto free = db().get_segment_manager()->get_free_memory();
+   const auto free = db().get_free_memory();
    const auto guard = my->conf.state_guard_size;
    EOS_ASSERT(free >= guard, database_guard_exception, "database free: ${f}, guard size: ${g}", ("f", free)("g",guard));
+
+   // give a change to chainbase to write some pages to disk if memory becomes scarce.
+   if (auto flushed_pages = mutable_db().check_memory_and_flush_if_needed()) {
```

Comment:
I think this call is only safe during the write window (at least, if it's actually clearing dirty bits). Are we sure `validate_db_available_size()` is only called during the write window? This one I'm not sure about due to ROtrx,
https://github.com/AntelopeIO/leap/blob/19f78f9b9a06f55a987e7d4c72f281b2772665ac/libraries/chain/controller.cpp#L2989-L2997

Reply:
Good catch. This is called from multiple read-only threads for read-only trx execution.

Reply:
Fixed.